src/java.base/share/classes/java/text/ListFormat.java

```diff
@@ -0,0 +1,600 @@
+    }
+
+    @Override
+    public AttributedCharacterIterator formatToCharacterIterator(Object arguments) {
+        Objects.requireNonNull(arguments);
+
+        if (arguments instanceof List<?> objs) {
+            var a = objs.toArray(new Object[0]);
+            return generateMessageFormat(a).formatToCharacterIterator(a);
+        } else if (arguments instanceof Object[] objs) {
+            return generateMessageFormat(objs).formatToCharacterIterator(objs);
+        } else {
+            throw new IllegalArgumentException("The arguments should be a List<Object> or an Object[]");
+        }
+    }
+
+    /**
+     * Checks if this {@code ListFormat} is equal to another {@code ListFormat}.
+     * The comparison is based on the {@code Locale} and formatting patterns, given or
+     * generated with {@code Locale}, {@code Type}, and {@code Style}.
+     * @param obj the object to check, {@code null} returns {@code false}
+     * @return {@code true} if this is equals to the other {@code ListFormat}
+     */
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+
+        if (obj instanceof ListFormat other) {
+            return locale.equals(other.locale) &&
+                Arrays.equals(patterns, other.patterns);
+        }
+
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode() {
+        return Objects.hash(locale, Arrays.hashCode(patterns));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
+        return
+            """
+            ListFormat [locale: "%s", start: "%s", middle: "%s",  end: "%s", two: "%s", three: "%s"]
+            """.formatted(locale.getDisplayName(), patterns[START], patterns[MIDDLE], patterns[END], patterns[TWO], patterns[THREE]);
+    }
+
+    private MessageFormat generateMessageFormat(Object[] input) {
+        var len = input.length;
+        return switch (len) {
+            case 0 -> throw new IllegalArgumentException("There should at least be one input string");
+            case 1 -> new MessageFormat("{0}", locale);
+            case 2, 3 -> new MessageFormat(patterns[len + 1], locale);
+            default -> new MessageFormat(createMessageFormatString(len), locale);
+        };
+    }
+
+    private String createMessageFormatString(int count) {
+        var sb = new StringBuilder(256).append(patterns[START]);
+        IntStream.range(2, count - 1).forEach(i -> sb.append(middleBetween).append("{").append(i).append("}"));
+        sb.append(patterns[END].replaceFirst("\\{0}", "").replaceFirst("\\{1}", "\\{" + (count - 1) + "\\}"));
+        return sb.toString();
+    }
+
+    @java.io.Serial
+    private void readObject(ObjectInputStream stream)
+            throws IOException, ClassNotFoundException {
+        stream.defaultReadObject();
+        try {
+            init();
+        } catch (IllegalArgumentException iae) {
```

Comment:
If the `patterns` array contains a null, perhaps due to corrupted stream contents, `init()` will throw NPE.
I don't recommend catching NPE here, but perhaps `init()` should check for nulls and throw IAE.

Reply:
Actually, null checks for the `patterns` array elements should be done also for the 1-arg factory method. Will add the check in `init()` and the corresponding test.